### PR TITLE
omnisharp-roslyn-mono: init at 1.37.17

### PR DESCRIPTION
**Pull request by elizagamedev**
-------

###### Description of changes

OmniSharp 1.38 drops support for dotnet 5.0, and consequently, mono.
This is a problem for Unity, which depends on a mono-compatible version
of OmniSharp for LSP.

This package is a slightly updated revision of the omnisharp-roslyn
package definition as it existed for 1.37.X releases.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by elizagamedev** [comment]
@tesq0 @ericdallo @corngood As maintainers of the original omnisharp-roslyn package, I thought it would be good to get your eyes on this. I apologize if it's not good etiquette to directly ping maintainers on PRs, I'm still new to the world of being a package maintainer.

**Comment by corngood** [comment]
Hey, another Unity gamedev using omnisharp-roslyn from nixpkgs!  I thought I was the only one.

My first concern with this is maintenance.  If it's at all possible to share some of it with `omnisharp-roslyn`, that would be great.  You'd probably need separate deps.nix, but most of the rest could probably be common.

My second concern that, as far as I can tell, there's no upstream maintenance of the 1.37 branch, so this will be a dead end.

That being said, I've been successfully using 1.38.2 on our Unity project for a while now.  I've been using emacs omnisharp-mode, rather than LSP, but I don't see why that would matter.  @elizagamedev what sort of problem were you having?

**Comment by ericdallo** [comment]
> My first concern with this is maintenance. If it's at all possible to share some of it with omnisharp-roslyn, that would be great. You'd probably need separate deps.nix, but most of the rest could probably be common.

Agreed, that would be nice

> That being said, I've been successfully using 1.38.2 on our Unity project for a while now. I've been using emacs omnisharp-mode, rather than LSP, but I don't see why that would matter. @elizagamedev what sort of problem were you having?

Dropping support for LSP sounds bad to me, LSP provides multiple features omnisharp-mode doesn't I think :(
Also, supporting LSP open doors for all other editors that have LSP support, not only Emacs 
(Emacs user with Unity here as well :) )

**Comment by ericdallo** [comment]
Sorry @corngood I misread you comment, makes sense to me

**Comment by corngood** [comment]
@corngood @elizagamedev this is pretty off topic, but since the Unity people are here... 

I have https://github.com/corngood/vscode-unity-debug, which I've been using with `dap-mode` to debug scripts in emacs.

I'm curious if you have any experience with debugging scripts in emacs.  I started a discussion https://github.com/corngood/vscode-unity-debug/discussions/1, but I'm not sure if it'll ping you.

**Comment by elizagamedev** [comment]
I just realized I may have made a pretty fundamental mistake re: mono support in most recent versions of omnisharp, but first, some replies:

> My first concern with this is maintenance. If it's at all possible to share some of it with `omnisharp-roslyn`, that would be great. You'd probably need separate deps.nix, but most of the rest could probably be common.

I agree, especially since mono is still supported unlike my initial assumption.

> On Linux I just use a shell script. As long as it starts with "rider", unity will use the rider IDE plugin. I wonder if that's why my projects work in 1.38.2?

I do something similar, but using the vscode IDE plugin (https://github.com/elizagamedev/unity.el). It sounds like the rider one might be better, especially if it handles csc.rsp. I will have to try it out and see if that works with my setup.

To elaborate a bit more about my setup: I'm using Emacs 29 (i.e. compiled on master branch) with eglot. Using the latest omnisharp-roslyn in nixpkgs 22.05 results in a deluge of "missing assembly" errors, but this is only a recent phenomenon that's been happening since the switch away from a mono build. I'm not sure what the fundamental reason is that we are experiencing such different behavior here is, maybe it's the rider vs vscode plugin (I'll have to test that). If you're using lsp-mode, I know that automatically downloads its own version of omnisharp; maybe that's what's happening?

> I'm curious if you have any experience with debugging scripts in emacs. I started a discussion [corngood/vscode-unity-debug#1](https://github.com/corngood/vscode-unity-debug/discussions/1), but I'm not sure if it'll ping you.

Unfortunately, I've been lucky enough not to ever have to use a C# debugger, so I can't offer much input.

Okay, so, yes, the latest omnisharp versions do still support mono builds, something that I completely missed. Basically, when investigating this issue, I stumbled upon [this comment](https://github.com/OmniSharp/omnisharp-roslyn/issues/1948#issuecomment-1038914951), which led me to realize that the switch recently happened in nixpkgs, which in turn led me to believe that the switch to a new-style project in 1.38 meant no more mono support. Instead of having a dedicated omnisharp-roslyn-mono definition file, we can probably add `mono` as a `bool` parameter to the derivation to build the package with mono support enabled, or something similar. Then we could define both omnisharp-roslyn and an omnisharp-rosyln-mono in all-packages.nix.

**Comment by corngood** [comment]
Unfortunately I can't access my dev machine to confirm right now, but I suspect you might need mono in your PATH for it to work?

I would try in the root of your unity project: `nix shell -p omnisharp-roslyn mono6 --run omnisharp`.  It should load all the projects without errors.

**Comment by elizagamedev** [comment]
> I would try in the root of your unity project: `nix shell -p omnisharp-roslyn mono6 --run omnisharp`. It should load all the projects without errors.

Nope, same errors, along the lines of:
```
{"Event":"Error","Body":{"Text":"System.IO.FileLoadException: Could not load file or assembly 'System.Configuration.ConfigurationManager...
```

Going to try with rider plugin-generated projects.

**Comment by elizagamedev** [comment]
> Going to try with rider plugin-generated projects.

No, this doesn't work either. I do suspect lsp-mode is either downloading and using a separate omnisharp binary for you, or otherwise changing your environment in some way.

**Comment by ericdallo** [comment]
Checking [lsp-mode's code](https://github.com/emacs-lsp/lsp-mode/blob/master/clients/lsp-csharp.el#L132-L139) it seems it expects to the server to be present on a path that was installed if you installed directly via lsp-mode with `lsp-install-server` which is not good, there is a way to use a specific server path though, via the `lsp-csharp-server-path` variable

**Comment by elizagamedev** [comment]
> I'm definitely using the omnisharp from nixpkgs. @elizagamedev could you possibly share a simple unity project which shows this error, along with generated solution/project files, and the full omnisharp log when you run it from nix-shell?

I actually realized that because I use flakes, I forgot to actually run `nix-channel --update` for a long time before running my `nix-shell` invocations in my previous comments, and so the errors and issues I'm encountering now are not 100% the same. So I did some proper digging. I am using:

- Unity version 2022.1.5f1 
- NixOS 22.05 + Flakes
- `nix-channel` tracking nixos-22.05
- Ran `nix-channel --update` beforehand
- So that I could run omnisharp on the CLI and in eglot with consistent configuration, I used a shell.nix file with direnv:

```nix
{ pkgs ? import <nixpkgs> { } }:

pkgs.mkShell {
  nativeBuildInputs = with pkgs; [
    omnisharp-roslyn
    mono6
  ];
}
```

To generate this test project, I did the following:

- Configured Unity's external editor to point to an executable called `code` which just does `exec "$@"`. This of course is done to convince Unity that it should generate a solution and project file unconditionally.
- Created a blank new project using the built-in 2D project template.
- Initialized an empty git repo in the project root so that eglot could detect the project root properly.
- Created a new behavior script with Create->C# Script, then double-clicked this file to ensure the solution/project files were generated.

Here is the full log:

```
{"Event":"log","Body":{"LogLevel":"INFORMATION","Name":"OmniSharp.Stdio.Host","Message":"Starting OmniSharp on Unknown 0.0 (Unknown)"},"Seq":1,"Type":"event"}{"Event":"log","Body":{"LogLevel":"INFORMATION","Name":"OmniSharp.Services.DotNetCliService","Message":"Checking the 'DOTNET_ROOT' environment variable to find a .NET SDK"},"Seq":2,"Type":"event"}
{"Event":"log","Body":{"LogLevel":"INFORMATION","Name":"OmniSharp.Services.DotNetCliService","Message":"DotNetPath set to /nix/store/ffza8aqvhv051lba17ab36l1xqhak5h6-dotnet-sdk-6.0.300/dotnet"},"Seq":3,"Type":"event"}
{"Event":"log","Body":{"LogLevel":"INFORMATION","Name":"OmniSharp.MSBuild.Discovery.MSBuildLocator","Message":"Located 1 MSBuild instance(s)\n            1: .NET Core SDK 6.0.300 17.2.0 - \"/nix/store/ffza8aqvhv051lba17ab36l1xqhak5h6-dotnet-sdk-6.0.300/sdk/6.0.300/\""},"Seq":4,"Type":"event"}
{"Event":"log","Body":{"LogLevel":"INFORMATION","Name":"OmniSharp.MSBuild.Discovery.MSBuildLocator","Message":"Registered MSBuild instance: .NET Core SDK 6.0.300 17.2.0 - \"/nix/store/ffza8aqvhv051lba17ab36l1xqhak5h6-dotnet-sdk-6.0.300/sdk/6.0.300/\""},"Seq":5,"Type":"event"}
{"Event":"log","Body":{"LogLevel":"INFORMATION","Name":"OmniSharp.WorkspaceInitializer","Message":"Invoking Workspace Options Provider: OmniSharp.Roslyn.CSharp.Services.CSharpFormattingWorkspaceOptionsProvider, Order: 0"},"Seq":6,"Type":"event"}
{"Event":"log","Body":{"LogLevel":"INFORMATION","Name":"OmniSharp.MSBuild.ProjectSystem","Message":"Detecting projects in '/home/eliza/Git/BrokenOmnisharpDemo/BrokenOmnisharpDemo.sln'."},"Seq":7,"Type":"event"}
{"Event":"log","Body":{"LogLevel":"INFORMATION","Name":"OmniSharp.MSBuild.ProjectManager","Message":"Queue project update for '/home/eliza/Git/BrokenOmnisharpDemo/Assembly-CSharp.csproj'"},"Seq":8,"Type":"event"}
{"Event":"log","Body":{"LogLevel":"INFORMATION","Name":"OmniSharp.Script.ScriptProjectSystem","Message":"Detecting CSX files in '/home/eliza/Git/BrokenOmnisharpDemo'."},"Seq":9,"Type":"event"}
{"Event":"log","Body":{"LogLevel":"INFORMATION","Name":"OmniSharp.Script.ScriptProjectSystem","Message":"Did not find any CSX files"},"Seq":10,"Type":"event"}{"Event":"log","Body":{"LogLevel":"INFORMATION","Name":"OmniSharp.WorkspaceInitializer","Message":"Configuration finished."},"Seq":11,"Type":"event"}
{"Event":"log","Body":{"LogLevel":"INFORMATION","Name":"OmniSharp.Stdio.Host","Message":"Omnisharp server running using Stdio at location '/home/eliza/Git/BrokenOmnisharpDemo' on host -1."},"Seq":12,"Type":"event"}
{"Event":"started","Body":null,"Seq":13,"Type":"event"}
{"Event":"log","Body":{"LogLevel":"INFORMATION","Name":"OmniSharp.MSBuild.ProjectManager","Message":"Loading project: /home/eliza/Git/BrokenOmnisharpDemo/Assembly-CSharp.csproj"},"Seq":14,"Type":"event"}
{"Event":"log","Body":{"LogLevel":"ERROR","Name":"OmniSharp.MSBuild.ProjectLoader","Message":"This project targets .NET version that requires reference assemblies that do not ship with OmniSharp out of the box (e.g. .NET Framework). The most common solution is to make sure Mono is installed on your machine (https://mono-project.com/download/) and that OmniSharp is started with that Mono installation (e.g. \"omnisharp.useGlobalMono\":\"always\" in C# Extension for VS Code)."},"Seq":15,"Type":"event"}
{"Event":"log","Body":{"LogLevel":"WARNING","Name":"OmniSharp.MSBuild.ProjectManager","Message":"Failed to load project file '/home/eliza/Git/BrokenOmnisharpDemo/Assembly-CSharp.csproj'."},"Seq":16,"Type":"event"}
{"Event":"MsBuildProjectDiagnostics","Body":{"FileName":"/home/eliza/Git/BrokenOmnisharpDemo/Assembly-CSharp.csproj","Warnings":[],"Errors":[{"LogLevel":"Error","FileName":"/nix/store/ffza8aqvhv051lba17ab36l1xqhak5h6-dotnet-sdk-6.0.300/sdk/6.0.300/Microsoft.Common.CurrentVersion.targets","Text":"This project targets .NET version that requires reference assemblies that do not ship with OmniSharp out of the box (e.g. .NET Framework). The most common solution is to make sure Mono is installed on your machine (https://mono-project.com/download/) and that OmniSharp is started with that Mono installation (e.g. \"omnisharp.useGlobalMono\":\"always\" in C# Extension for VS Code).","StartLine":1221,"StartColumn":5,"EndLine":0,"EndColumn":0}]},"Seq":17,"Type":"event"}
{"Event":"log","Body":{"LogLevel":"ERROR","Name":"OmniSharp.MSBuild.ProjectManager","Message":"Attempted to update project that is not loaded: /home/eliza/Git/BrokenOmnisharpDemo/Assembly-CSharp.csproj"},"Seq":18,"Type":"event"}
```

As you can see, there are a lot of errors about the .NET target, with a recommendation to use the "global mono" if you're using VS Code. I recall having issues with this when I used lsp-mode in emacs, since lsp-mode would automatically download a "non-global mono" version of omnisharp which didn't work with Unity projects unless you messed around with some environment variables.

So, I referred back to a blog post I wrote about this subject [from before I was using NixOS](https://eliza.sh/2021-06-01-using-unity-editor-with-emacs.html), and set the `FrameworkPathOverride` environment variable in shell.nix.

```nix
{ pkgs ? import <nixpkgs> { } }:

pkgs.mkShell {
  nativeBuildInputs = with pkgs; [
    omnisharp-roslyn
    mono6
  ];

  shellHook = ''
    export FrameworkPathOverride="${pkgs.mono6}/lib/mono/4.5"
  '';
}
```

And... it fucking works perfectly. 😒

So it seems like the mono version of omnisharp worked out of the box for me in Emacs for a long while, but the switch to the non-mono version brought up this edge-case which completely broke it. Now I'm wondering why this wasn't an issue for you and if there's any way to change the omnisharp or mono package so that this really awful hack isn't necessary. Maybe it's because I don't have mono6 installed in my `home.packages` or my `environment.systemPackages`?

In conclusion: I am so, so tired of dealing with C# and .NET. 🫤

**Comment by elizagamedev** [comment]
Yep, the rider plugin actually does generate working projects. I did test it before, but that was on an older version of omnisharp. I guess my issue all along was that I was using an old broken version that didn't work with anything.

Thank you for your help with figuring out this issue. It seems like most of the conventional wisdom I could find about this subject was just straight-up wrong. I'll close this issue now and incorporate the solution here into unity.el and probably write another blog post.

**Comment by corngood** [comment]
Ah, that's great.  IMO the IDE plugins are a mess in Unity right now.  There are three of them that all generate project files in the same format but without any shared code :|.  I'd toyed with the idea of forking the rider plugin to make something that would work out of the box with other omnisharp tools, but so far the `rider` shim has been manageable.  There's also no way to install per -user packages afaik, which makes it cumbersome to add new IDE plugins.
